src/components/lib/posix/posix.c

```diff
@@ -293,6 +298,152 @@ cos_clone(int (*func)(void *), void *stack, int flags, void *arg, pid_t *ptid, v
 	return thd->thdid;
 }
 
+#define FUTEX_WAIT		0
+#define FUTEX_WAKE		1
+#define FUTEX_FD		2
+#define FUTEX_REQUEUE		3
+#define FUTEX_CMP_REQUEUE	4
+#define FUTEX_WAKE_OP		5
+#define FUTEX_LOCK_PI		6
+#define FUTEX_UNLOCK_PI		7
+#define FUTEX_TRYLOCK_PI	8
+#define FUTEX_WAIT_BITSET	9
+
+#define FUTEX_PRIVATE 128
+
+#define FUTEX_CLOCK_REALTIME 256
+
+struct futex_data
+{
+	int *uaddr;
+	struct ps_list_head waiters;
+};
+
+struct futex_waiter
+{
+	thdid_t thdid;
+	struct ps_list list;
+};
+
+#define FUTEX_COUNT 20
+struct futex_data futexes[FUTEX_COUNT];
+
+struct futex_data *
+lookup_futex(int* uaddr)
+{
+	int last_free = -1;
+	int i;
+
+	for (i = 0; i < FUTEX_COUNT; i++) {
+		if (futexes[i].uaddr == uaddr) {
+			return &futexes[i];
+		} else if (futexes[i].uaddr == 0) {
+			last_free = i;
+		}
+	}
+	if (last_free >= 0) {
+		futexes[last_free] = (struct futex_data) {
+			.uaddr = uaddr
+		};
+		ps_list_head_init(&futexes[last_free].waiters);
+		return &futexes[last_free];
+	}
+	printc("Out of futex ids!");
+	assert(0);
+}
+
+/* TODO: Cleanup empty futexes */
+
+struct sl_lock futex_lock = SL_LOCK_STATIC_INIT();
+
+int
+cos_futex_wait(struct futex_data *futex, int *uaddr, int val, const struct timespec *timeout)
+{
+	/* We enter the function with futex_lock already taken */
+	int not_awoken;
+	cycles_t   deadline;
+	microsec_t wait_time       = 0;
+	struct futex_waiter waiter = (struct futex_waiter) {
+		.thdid = sl_thdid()
+	};
+
+	if (*uaddr != val) return EAGAIN;
+
+	ps_list_init_d(&waiter);
+	ps_list_head_append_d(&futex->waiters, &waiter);
+
+	if (timeout != NULL) {
+		wait_time = time_to_microsec(timeout);
+	}
+
+	/* No race here, we'll enter the awoken state if things go wrong */
+	sl_lock_release(&futex_lock);
+	if (wait_time == 0) {
+		sl_thd_block(0);
+	} else {
+		deadline = sl_now() + sl_usec2cyc(wait_time);
+		sl_thd_block_timeout(0, deadline);
+	}
+	sl_lock_take(&futex_lock);
+	ps_list_rem_d(&waiter);
```

Comment:
`waiter` should likely have been removed in `wake` rather than here.  If it is not removed in `wake`, we can have multiple wakes for the same thread which is not really what we want.